src/alarms/alarm-handling.md

```diff
@@ -0,0 +1,87 @@
+
+A lost alam-raise could be even more problematic if the raised alarm requires some cloud-site interaction (or manual interaction). Since the lost alarm is not visible on cloud no intercation will be started. And since not state-change happen again, the alarm will never appear at cloud.
+
+**Consequence:**<br/>
+All alarm state-changes need to be transfered reliable from **Application to thin-edge**, and from **thin-edge to cloud**.
+I.E. alarm state-changes must not got lost, or device software must be able to detect loss and react accordingly (e.g. retry to transfer).
+
+More details about loss off data-transfer and proper reaction see...  TODO: add data-flow-image here!
+
+# Information Set per Alarm
+The specified alarm attributes below are inspired by cumulocity data model for alarms, but shall be re-usable for other clouds where possible.
+
+| Name              | Description   |
+| ----------------- |---------------|
+| ref-string        | Device-unique reference string (kind of alarm ID or alarm name), used to reference the once occured alarm again, e.g. in case of updates (e.g. "temperature_sensor_loss") |
+| message-string    | Human readable short information about alarm reason (e.g. "Temperature sensor does not respond") |
+| severity-string   | Could be "CRITICAL", "MAJOR", "MINOR" or "WARNING" |
+| status-string     | Could be "ACTIVE" or "CLEARED"<br/>
+| time-string       | Timestamp indicating when the alarm (or the alarm-update) had occured (in [ISO 8601 format](http://www.w3.org/TR/NOTE-datetime#)) |
+
+# Data flow
+
+Figure below illustrates the data flow from Customer Application broker/thin-edge up to cloud. It shows especially the use of MQTT QOS=1, MQTT Retain and Mosquitto Persistence to achieve reliability alarm state-change transfer.
+
+![Sequence Diagram Update SW-list](images/alarm_dataflow.svg)
+
+**TODO:** Interface from Mapper to Cloud to be defined. Two options are possible:
+1) **JSONviaMQTT:** To have reliability use QOS=1. Open issue: Not yet completely confirmed that C8Y sends PUBACK when msg was processed, instead just on arrival.
+2) **HTTP REST:** HTTP response indicates if message was processed sucessfully. Anyway HTTP REST would increase complexity of implementation and increase data traffic.
+
+
+# Public MQTT-based alarm interface
+
+Similar to thin-edge'S measurement interface, alarm interface is based on MQTT topics.
+
+## Topic structure and payload
+
+Proposal to have "severity" and "ref-string" as topics:
+
+**thin-edge JSON (alarm) format:**
+```
+Topic:   tedge/alarms/<device-id>/<severity-string>/<ref-string>
+Payload: {
+  "message":  <message-string>,
+  "status":   <status-string>,  
+  "time":     <time-string>
+}
+```
+
+NOTE: The "device-id" indicates whether it is an alarm of the thin-edge device it-self (root-device) or a child-device. Use "root" for the thin-edge device or the child-device id for the according child-device.

```

Comment:
What if the child-device id is `root`?

Reply:
I suppose it also would be good if other tedge topics are consistent with alarms API and therefore having `tedge/measurements` to use similar structure would be good (if not now at least before 1.0 release).

I suppose instead of this being `root` we could use `device name` from certificate (but then we need to have a way to get that name if we don't have cert or cert is invalid).

Reply:
Also, what happens if device names have a `/` in them?

Reply:
An option to avoid ambiguities is to do as for the `tegde/measurements`.

* Alarms on the device are raised on `tedge/alarms/<severity-string>/<ref-string>`
* Alarms on a child device are raised on `tedge/alarms/<severity-string>/<ref-string>/<device-id>`

Reply:
> I suppose it also would be good if other tedge topics are consistent with alarms API and therefore having tedge/measurements to use similar structure would be good (if not now at least before 1.0 release).

That's right. My personal plan is to bring `tedge/measurements` into a similar topic-structure before release 1.0.

Reply:
> I suppose instead of this being root we could use device name from certificate (but then we need to have a way to get that name if we don't have cert or cert is invalid).

I prefer to have some constant identifier to address the device. Dynamic things complicate implementation for the customer, and increases risk for bugs at customer side (e.g. when a customer hard-codes that device-id, it works on the desk, but the problem occurs not before role-out to other devices.).

Reply:
> An option to avoid ambiguities is to do as for the tegde/measurements.
> 
>     Alarms on the device are raised on tedge/alarms/<severity-string>/<ref-string>
>     Alarms on a child device are raised on tedge/alarms/<severity-string>/<ref-string>/<device-id>

Technical I see just one small downside: It might not allow easily backward compatible future API extensions.

Example:
Todays topic:
`tedge/alarms/<severity-string>/<ref-string>/<device-id>`
If we want to introduce a some `new-attribute` for the alarm API:
OPTION 1:
`tedge/alarms/<severity-string>/<ref-string>/<new-attrib>/<device-id>`
-> That would break the API compatibility.
OPTION 2:
`tedge/alarms/<severity-string>/<ref-string>/<device-id>/<new-attrib>`
-> With that the `device-id` could clash again with root-device's new attrib, which is at: `tedge/alarms/<severity-string>/<ref-string>/<new-attrib>`

My proposal: Move all child device alarms just underneath the topic `childs`, as below:

For root-device:
`tedge/alarms/<severity-string>/<ref-string>/`
For childs:
`tedge/alarms/<severity-string>/childs/<child-id>/<ref-string>/`

Examples:
`tedge/alarms/major/alarm1`
`tedge/alarms/major/childs/child1/alarm1`
`tedge/alarms/major/childs/child2/alarm2`
`tedge/alarms/major/childs/child3/alarm4711`

That implies that `childs` can not be used as alarm ref-string, but it's just a matter of definition in context of thin-edge.